### DISCUSSION TITLE
Exporter omits log formatter

### DISCUSSION THREAD

**Discussion by jose-mf** [open-telemetry/opentelemetry-dotnet#4020] [Q&A]
Hi there, I don't know if I miss something in the configuration but when I try to use a log formatter always it's omitted and the output format is a simple string.

This is the example code using a ConsoleExporter:

 var serviceName = "Samples.ConsoleExporter";
ActivitySource source;

using var tracerProvider = Sdk.CreateTracerProviderBuilder()
                    .AddSource(serviceName)
                    .AddConsoleExporter()
                    .Build();

if (Activity.Current == null)
{
    source = new ActivitySource(serviceName);
}
else
{
    source = Activity.Current.Source;
}

using var loggerFactory = LoggerFactory.Create(builder =>
{
    builder.ClearProviders();
    builder.AddJsonConsole();
    builder.AddOpenTelemetry(opt =>
    {
        opt.AddConsoleExporter();
        opt.IncludeFormattedMessage = true;
    });
});

using var rootSpan = source.StartActivity("Samples.ConsoleExporter.Handler", ActivityKind.Server);

var logger = loggerFactory.CreateLogger<Program>();

logger.LogInformation("Traces are being created and exported");

And the output is:

LogRecord.Timestamp:               2022-12-16T09:20:58.2482340Z
LogRecord.TraceId:                 69e3c44e83459b5cbc6d71adb837afea
LogRecord.SpanId:                  057c72070fc3bdc1
LogRecord.TraceFlags:              Recorded
LogRecord.CategoryName:            Program
LogRecord.LogLevel:                Information
LogRecord.FormattedMessage:        Traces are being created and exported
LogRecord.State (Key:Value):
    OriginalFormat (a.k.a Body): Traces are being created and exported

Resource associated with LogRecord:
service.name: unknown_service:dotnet

Activity.TraceId:            69e3c44e83459b5cbc6d71adb837afea
Activity.SpanId:             057c72070fc3bdc1
Activity.TraceFlags:         Recorded
Activity.ActivitySourceName: Samples.ConsoleExporter

The expected log format should be:
{"EventId":0,"LogLevel":"Information","Category":"Program","Message":"Traces are being created and exported","State":{"Message":"Traces are being created and exported","{OriginalFormat}":"Traces are being created and exported"}}

**Comment by cijothomas** [comment]
> I try to use a log formatter always it's omitted and the output format is a simple string.

Could you elaborate? I am not sure if I follow the issue. LogRecord.FormattedMessage contains the formatted message as expected. What is missing?

**Reply by jose-mf** [reply]
You can see in the output that the LogRecord.FormattedMessage and the OriginalFormat contain the same data. It seems the option "includeFormattedMessage" is ignored somehow.
If you execute the code I pasted in the original message you can reproduce the behavior I'm describing.

The expected content in the LogRecord.FormattedMessage should be:

{"EventId":0,"LogLevel":"Information","Category":"Program","Message":"Traces are being created and exported","State":{"Message":"Traces are being created and exported","{OriginalFormat}":"Traces are being created and exported"}}

**Reply by cijothomas** [reply]
The example uses unstructured logging (in this case a simple string), so the original format and formatted message is same - just the string.

For the logging statement which can be one way of doing structured logging, _"logger.LogInformation("Hello from {name} {price}.", "tomato", 2.99)",_, the Body will be "Hello from {name} {price}" by default. If `IncludeFormattedMessage` is true, then the body will be "Hello from tomato 2.99". In both cases, name, price will be printed as key-value pairs too.

Under the hoods, the `IncludeFormattedMessage` decides if body will be populated from "template" vs the "formattedMessage" which is the result of calling the "formatter" from the Log call. (https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.ilogger.log?view=dotnet-plat-ext-7.0)

Let me know if this clarifies.

**Reply by jose-mf** [reply]
I didn't understand well the behavior of IncludeFormattedMessage. I thought it was related if the output log message was a simple string or a structured JSON message. Then, Let me rephrase the question:
If I have a logging provider like **AddJsonConsole** and the output format is in a JSON format, I should see the JSON value in the OriginalFormat property. Or Why is the reason I only see the content of the _logger.LogInformation("Hello World")_, for example.

I hope it clarifies more my question/problem.

**Reply by cijothomas** [reply]
`IncludeFormattedMessage` only determines if we invoke the `formatter`, which is the last parameter to ILogger.Log call.
It is unrelated to the AddJsonConsoles formatting, which is mostly about how the output is formatted while displaying on console.

See also : https://github.com/open-telemetry/opentelemetry-dotnet/issues/2042#issuecomment-1098588590
